### PR TITLE
Pull req same jms

### PR DESCRIPTION
**Pull request by ghost**
This is Solace JMS API
Changed to command line per our conversation

### CODE REVIEW

**Comment by ghost** [comment]
Jonathan, it' s all yours

File: src/main/java/com/solace/samples/TopicSubscriber.java [hunk 1]
```diff
@@ -34,79 +37,85 @@
...
-            latch.await(); // block here until message received, and latch will
-                           // flip
-        } catch (InterruptedException e) {
-            System.out.println("I was awoken while waiting");
-        }
-
-        // Close consumer
+        System.out.println("Awaiting message...");
+        // the main thread blocks at the next statement until a message received
+        latch.await();
```
Comment (mdspielman) [inline src/main/java/com/solace/samples/TopicSubscriber.java:102]: This code reformatting moves this code away from the associated markdown tutorial. If proposing this change, please also update the markdown tutorial as well. It also moves this code away from other samples in this project which follow a similar code pattern. Can we please keep the samples internally consistent.

File: src/main/java/com/solace/samples/TopicPublisher.java [hunk 2]
```diff
@@ -26,57 +30,74 @@
...
 
+/**
+ * Publishes a messages to a topic using Solace JMS 1.1 API implementation.
+ * 
+ * This is the Publisher in the Publish/Subscribe messaging pattern.
+ */
 public class TopicPublisher {
 
-    public void run(String... args) throws Exception {
+    final String SOLACE_VPN = "default";
```
Comment (mdspielman) [inline src/main/java/com/solace/samples/TopicPublisher.java:44]: This change, moves these two samples away from the rest of the samples in this JMS project (and the style of the rest of the Solace samples) in how the VPN, username and password are handled. At a minimum, I would like to keep the JMS samples internally consistent. So if we change the way arguments are handled in these two samples, I would like to make sure the other JMS samples and associated markdown tutorials are also updated in this pull request.

**Review by mdspielman — CHANGES_REQUESTED** [review]
Jonathan will do a more complete review, but here are some initial thoughts on the pull request.

File: src/main/java/com/solace/samples/TopicPublisher.java (hunk 2 above)
Comment (jschabowsky) [inline src/main/java/com/solace/samples/TopicPublisher.java:44]: Please remove the setting of variables in the code and revert back to command line for now. (as you did for Qpid)

File: src/main/java/com/solace/samples/TopicSubscriber.java [hunk 3]
```diff
@@ -34,79 +37,85 @@
...
+/**
+ * Subscribes to messages published to a topic using Solace JMS 1.1 API implementation.
+ *
+ * This is the Subscriber in the Publish/Subscribe messaging pattern.
+ */
 public class TopicSubscriber {
 
-    public void run(String... args) throws Exception, JMSException, NamingException {
-        System.out.println("TopicSubscriber initializing...");
+    final String SOLACE_VPN = "default";
```
Comment (jschabowsky) [inline src/main/java/com/solace/samples/TopicSubscriber.java:46]: Same comment as before. revert back to the command line args

**Review by jschabowsky — CHANGES_REQUESTED** [review]
Thanks! 

Just revert back to command line args and follow up on marks comments and I think I am good!

File: src/main/java/com/solace/samples/TopicSubscriber.java (hunk 1 above)
Comment (ghost) [reply src/main/java/com/solace/samples/TopicSubscriber.java:102]: Mark, of course the tutorial will change as I repeatedly stated it in related email threads. But first I need to make sure the code is accepted.

File: src/main/java/com/solace/samples/TopicPublisher.java (hunk 2 above)
Comment (ghost) [reply src/main/java/com/solace/samples/TopicPublisher.java:44]: Mark, how do want it to be handled? I copied the existing argument code as is. Please be specific.

**Comment by ghost** [comment]
Guys, this pull request consists of TWO commits, you reviewed the first one, but not the second one that has the command line argument processing in that ugly way you wish.

image

**Comment by jschabowsky** [comment]
Oh, Sorry.

Can you send me a link to that pull req?

**Comment by ghost** [comment]
Jonathan, this is the pull request. We're on the "Conversation" tab, the next one is "Commits" that lists both commits I'm talking about

**Comment by ghost** [comment]
Now includes the tutorial changes

File: src/main/java/com/solace/samples/TopicSubscriber.java (hunk 1 above)
Comment (ghost) [reply src/main/java/com/solace/samples/TopicSubscriber.java:102]: see commit dcc1334  https://github.com/SolaceSamples/solace-samples-jms/pull/15/commits

**Review by jschabowsky — APPROVED** [review]
Looks good

**Comment by ghost** [comment]
Mark, it's ready for merging in. See commit dcc13347 for tutorial text change.

**Comment by mdspielman** [comment]
As per offline conversation, code is moving in the right direction towards merging. To finish this pull request please address the following:

* Markdown tutorial code snippets and sample code should be copies of each other. Example of where it is not is connection factory creation in TopicPublisher.
* If choosing to make these style changes that you have proposed in this pull request, please reflect them in all tutorials and associated markdown tutorials within the project. I'd like all tutorials to updated in one pull request to keep the style uniform.

**Review by mdspielman — APPROVED** [review]

**Comment by ghost** [comment]
Changed two other tutorials code to mimic the previously sorted out pub/sub changes.

**Comment by ghost** [comment]
Now it includes 5 commits

**Comment by mdspielman** [comment]
@dfedorov-solace, can you please resolve the branch conflicts and then we'll organize a review.

**Comment by ghost** [comment]
Yes, I'm in the middle of working on it

**Comment by ghost** [comment]
mdspielman Mark, I think it's complete now, please review, thank you for your patience.
BTW, do you want me to squish all commits so it's easier for you to look at?

**Comment by ghost** [comment]
I see the build error, fixing

**Comment by ghost** [comment]
all good now

**Comment by ghost** [comment]
Mark, I'm all done with this project - all runs against it were successful, just a one small check for corrected, the project is ready for your review and merge.

**Comment by mdspielman** [comment]
@jschabowsky, please do a new code review.

File: docs/_tutorials/persistence-with-queues.md [hunk 4]
```diff
@@ -119,11 +120,7 @@ To send a message, you must still create a message. The difference from sending
 
 ```java
 TextMessage message = session.createTextMessage("Hello world Queues!");
-producer.send(queue,
-              message,
-              DeliveryMode.PERSISTENT,
-              Message.DEFAULT_PRIORITY,
-              Message.DEFAULT_TIME_TO_LIVE);
+messageProducer.send(message, DeliveryMode.PERSISTENT, Message.DEFAULT_PRIORITY, Message.DEFAULT_TIME_TO_LIVE);
```
Comment (mdspielman) [inline docs/_tutorials/persistence-with-queues.md]: Can you revert to passing the destination in the call to send(). I believe it is more clear for the samples and matches the discussion in the associated tutorials.

**Review by mdspielman — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
Changed in https://github.com/SolaceSamples/solace-samples-jms/pull/15/commits/6ca135a3016b826085de364febbaa5d600112dca

**Comment by jschabowsky** [comment]
Looks good! Thanks!

**Comment by jschabowsky** [comment]
Ok, I may have spoken too soon. What we need to do, is match the existing JMS Samples on how you get the API (Section: Obtaining JMS 1.1 API). So take a look at the existing Solace-JMS-Examples and match that... (In other words make it easy to pull in qpid)... Then in the section: "Building" match the existing as well.

**Review by mdspielman — APPROVED** [review]

File: docs/_tutorials/persistence-with-queues.md [hunk 5]
```diff
@@ -137,29 +134,32 @@ Now it is time to receive the messages sent to your queue.
 You still need a JMS `Connection` just as you did with the producer. With a connection, you then need to create a Session and bind to the Solace message router queue by creating a `MessageConsumer`. This is nearly identical to what was shown in the publish/subscribe tutorial. In this case, create a Session but use the Solace client acknowledgement mode. This allows the consumers to acknowledge each message individually without side-effects. You can learn more about acknowledgement modes in the Establishing Connections sections of [Solace JMS Documentation – Establishing Connections]({{ site.docs-jms-connections }}){:target="_top"}.
 
 ```java
-Session session = connection.createQueueSession(false, SupportedProperty.SOL_CLIENT_ACKNOWLEDGE);
+Session session = connection.createSession(false, Session.CLIENT_ACKNOWLEDGE);
```
Comment (mdspielman) [inline docs/_tutorials/persistence-with-queues.md]: Please revert. Using SOL_CLIENT_ACKNOWLEDGE is the preferred way to ack a message with Solace JMS.

File: docs/_tutorials/request-reply.md [hunk 6]
```diff
@@ -202,24 +197,23 @@ All that’s left is to receive and process the reply message as it is received
...
-
-if (reply instanceof TextMessage) {
-    System.out.printf("TextMessage response received: '%s'%n",
-        ((TextMessage)reply).getText());
-    if     (!reply.getBooleanProperty(SupportedProperty.SOLACE_JMS_PROP_IS_REPLY_MESSAGE)) {
-
-        System.out.println("Warning: Received a reply message without the isReplyMsg flag set.");
+    System.out.println(
+            "Received a reply message with no correlationID. This field is needed for a direct request.");
+} else {
```
Comment (mdspielman) [inline docs/_tutorials/request-reply.md]: Please revert this code. You've introduced a null pointer exception.

File: docs/_tutorials/using-jndi.md [hunk 7]
```diff
@@ -206,8 +213,8 @@ The [Solace JMS Documentation - JMS Properties Reference]({{site.docs-jms-proper
 Next, the 'ConnectionFactory' can be used the same way as described in the Persistence with Queues tutorial to create a JMS Connection, at which point your client is connected to the Solace message router and can create a JMS Session.
 
 ```java
-Connection connection = cf.createConnection();
-final Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+Connection connection = connectionFactory.createConnection();
+final Session session = connection.createSession(false, Session.CLIENT_ACKNOWLEDGE);
```
Comment (mdspielman) [inline docs/_tutorials/using-jndi.md]: Please revert or align with the SOL_CLIENT_ACKNOWLEDGE used in the receivers.

File: src/main/java/com/solace/samples/BasicRequestor.java [hunk 8]
```diff
@@ -34,98 +38,115 @@
...
+                } else {
+                    if (reply instanceof TextMessage) {
+                        System.out.printf("TextMessage response received: '%s'%n", ((TextMessage) reply).getText());
+                    } else {
+                        System.out.println("Message response received.");
+                    }
+                    if (!reply.getBooleanProperty(SupportedProperty.SOLACE_JMS_PROP_IS_REPLY_MESSAGE)) {
+                        System.out.println("Warning: Received a reply message without the isReplyMsg flag set.");
+                    }
+                    System.out.printf("Message Content:%n%s%n", SolJmsUtility.dumpMessage(reply));
```
Comment (mdspielman) [inline src/main/java/com/solace/samples/BasicRequestor.java]: This printout has been made redundant by the above code. Suggest removing it or reverting the code.

File: src/main/java/com/solace/samples/QueueConsumer.java [hunk 9]
```diff
@@ -17,109 +17,117 @@
...
-        cf.setDynamicDurables(true);
+        connectionFactory.setDynamicDurables(true);
+
+        // Create connection to the Solace router
+        Connection connection = connectionFactory.createConnection();
 
-        // JMS Connection
-        QueueConnection connection = cf.createQueueConnection();
+        // Create a non-transacted, client ACK session.
+        Session session = connection.createSession(false, Session.CLIENT_ACKNOWLEDGE);
```
Comment (mdspielman) [inline src/main/java/com/solace/samples/QueueConsumer.java]: Please revert to SOL_CLIENT_ACKNOWLEDGE here too.

File: src/main/java/com/solace/samples/QueueConsumerJNDI.java [hunk 10]
```diff
@@ -17,111 +17,128 @@
...
 
-        // Create a non-transacted, Client Ack session.
-        Session session = connection.createQueueSession(false, SupportedProperty.SOL_CLIENT_ACKNOWLEDGE);
+        // Create connection to the Solace router
+        Connection connection = connectionFactory.createConnection();
 
-        // Lookup Queue.
-        Queue queue = (Queue) initialContext.lookup("/JNDI/Q/tutorial");
+        // Create a non-transacted, client ACK session.
+        Session session = connection.createSession(false, Session.CLIENT_ACKNOWLEDGE);
```
Comment (mdspielman) [inline src/main/java/com/solace/samples/QueueConsumerJNDI.java]: Revert to SOL_CLIENT_ACKNOWLEDGE

**Review by mdspielman — CHANGES_REQUESTED** [review]
On final review, I noticed a few more errors. If you could please address then I'll merge.

File: docs/_tutorials/persistence-with-queues.md (hunk 5 above)
Comment (mdspielman) [reply docs/_tutorials/persistence-with-queues.md]: An alternative approach would be to keep it as CLIENT_ACK and add a link to this explanation which shows a Solace alternative which solves a specific application problem with the specification version of the JMS ack: http://docs.solace.com/API-Developer-Online-Ref-Documentation/jms/com/solacesystems/jms/SupportedProperty.html#SOL_CLIENT_ACKNOWLEDGE

Comment (ghost) [reply docs/_tutorials/persistence-with-queues.md]: I'll do the preferred way.

File: docs/_tutorials/using-jndi.md (hunk 7 above)
Comment (ghost) [reply docs/_tutorials/using-jndi.md]: done

File: src/main/java/com/solace/samples/QueueConsumer.java (hunk 9 above)
Comment (ghost) [reply src/main/java/com/solace/samples/QueueConsumer.java]: done

File: src/main/java/com/solace/samples/QueueConsumerJNDI.java (hunk 10 above)
Comment (ghost) [reply src/main/java/com/solace/samples/QueueConsumerJNDI.java]: done

File: src/main/java/com/solace/samples/BasicRequestor.java (hunk 8 above)
Comment (ghost) [reply src/main/java/com/solace/samples/BasicRequestor.java]: Mark, I'm not sure I understand how you got to this conclusion, I'd suggest we look at it tomorrow morning together, please drop by at the time of your convenience.

File: docs/_tutorials/request-reply.md (hunk 6 above)
Comment (ghost) [reply docs/_tutorials/request-reply.md]: Mark, I'm not sure I understand how you got to this conclusion, I'd suggest we look at it tomorrow morning together, please drop by at the time of your convenience.

Comment (mdspielman) [reply docs/_tutorials/request-reply.md]: You're correct. I made a mistake. There is not a null pointer issue. Personally I prefer the original style but will accept this modification if you prefer it.

**Comment by ghost** [comment]
@mdspielman  Mark, see 17541686a6d391dee4ef8135563dcd43e3a885fc for the changes you requested.

About the pinetree, I finally recalled why I changed it to ifs instead of exceptions: the stop()/close() routines at the end of that routine that need to be called for a graceful exit.

Let's talk about it in the morning.

**Comment by ghost** [comment]
@mdspielman see 7b51ba64f509dbef5318b9406c8a18f879cf4070

File: docs/_tutorials/request-reply.md [hunk 11]
```diff
@@ -202,24 +197,23 @@ All that’s left is to receive and process the reply message as it is received
...
+    if (!reply.getJMSCorrelationID().equals(correlationId)) {
```
Comment (mdspielman) [inline docs/_tutorials/request-reply.md]: This code block should be reverted I think?

File: docs/_tutorials/using-jndi.md [hunk 12]
```diff
@@ -217,9 +224,11 @@ A Queue or Topic destination is needed to send and receive messages. When using
 Following code will obtain a Queue for the Persistence with Queues scenario:
 
 ```java
-Queue queue = (Queue)initialContext.lookup("/JNDI/Q/tutorial");
+final String QUEUE_NAME = "Q/tutorial";
+final String QUEUE_JNDI_NAME = "/JNDI/" + QUEUE_NAME;
+Queue queue = (Queue) initialContext.lookup(QUEUE_JNDI_NAME);
 ```
-
+SupportedProperty.SOL_CLIENT_ACKNOWLEDGE
```
Comment (mdspielman) [inline docs/_tutorials/using-jndi.md]: This seems out of place.

**Comment by ghost** [comment]
@mdspielman see 72cf5d95a0fe24a7a4920d6e6ae0420c88ae09b8

**Comment by mdspielman** [comment]
Changes look complete. Merging.

**Review by mdspielman — APPROVED** [review]

**Pull request merged**